examples/hsm-example.cpp

```diff
@@ -26,20 +28,94 @@
+        /* The argument hashFunction is optional. Default is SHA256
+         * The default signature format is ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS */
+        signCtx = std::make_shared<ECDSASignaturePrivateKeyCtx>(privKey, digestType, sigFormat);
+    } catch (MoCOCrWException &e) {
+        std::cerr << "Please check your ECC key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+
+    std::vector<uint8_t> signature;
+    try {
+        signature = signCtx->signMessage(message);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - error in openssl (sign, padding, ...)
+         * - Hash function's digest size doesn't match the message's digest size
+         * - Invalid signature format set
+         */
+        std::cerr << "Failure occurred during signing." << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+    return signature;
+}
+
+void ecdsaVerify(const AsymmetricPublicKey &pubKey,
+                 const DigestTypes digestType,
+                 const ECDSASignatureFormat sigFormat,
+                 const std::vector<uint8_t> &signature,
+                 const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageVerificationCtx> verifyCtx;
+    try {
+        verifyCtx = std::make_shared<ECDSASignaturePublicKeyCtx>(pubKey, digestType, sigFormat);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Please check your ECC key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+    try {
+        verifyCtx->verifyMessage(signature, message);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - error in openssl (sign, padding, ...)
+         * - Hash function's digest size doesn't match the message's digest size
+         * - Invalid signature format set
+         * - Signature can't be parsed
+         * - Invalid signature
+         */
+        std::cerr << "Verification failed!" << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
 int main(void)
 {
     // Information for engine loading and key management.
     std::string id("pkcs11");
     std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
     std::string pin("1234");
     std::string keyID("1001");
-
-    // Step 2: Initialise the PKCS11 Engine.
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
     HsmEngine hsmEngine(id, modulePath, pin);
 
-    // Step 4: Load the key from the HSM and print it.
-    auto pubKey = AsymmetricPublicKey::readPublicKeyFromHSM(hsmEngine, keyID);
-    std::cout << "Loaded key: \n\n" << pubKey.publicKeyToPem() << "\n\n" << std::endl;
+    /************** ECDSA signature **************/
+    auto ecdsaDigestType = DigestTypes::SHA512;
+    ECCSpec ecspec;
+    auto eccPrivKey = AsymmetricPrivateKey::genKeyOnHsmGetPrivate(
```

Comment:
In addition, also include an RSA example for completeness? (I know EC is more important, but would be nice to have both, especially since these examples also serve as additional tests).